Bug 1864243 - Use alsa plug to interact with MIDI devices

ebababi opened the pull request:
Bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1864243

Firefox implementation of WebMIDI API requires ALSA in order to interact with MIDI devices.

ebababi committed:
Bug 1864243 - Use alsa plug to interact with MIDI devices

Firefox implementation of WebMIDI API requires ALSA in order to interact with MIDI devices.

lissyx commented:
@bandali0

ebababi commented:
Btw, you may want to try ALSA virtual MIDI devices to test this:
```
$ sudo modprobe snd-virmidi snd_index=1
$ amidi -l                             
Dir Device    Name
IO  hw:4,0    Virtual Raw MIDI (16 subdevices)
IO  hw:4,1    Virtual Raw MIDI (16 subdevices)
IO  hw:4,2    Virtual Raw MIDI (16 subdevices)
IO  hw:4,3    Virtual Raw MIDI (16 subdevices)
```

To be honest I'm not very familiar with `snapcraft` and I couldn't really test this. Based on [this post](https://forum.snapcraft.io/t/the-alsa-interface/7766/7), there might be one more step that we need to take in order for this to work. Happy to test any builds with real hardware.

seb128 commented:
Could you gives details on 'requires ALSA'? What exactly in alsa? Does it call into command line utilities?

I've rebuilt a snap with the proposed change and it doesn't make a difference (trying the example code from https://www.smashingmagazine.com/2018/03/web-midi-api/), also the journal doesn't show any apparmor denial which it probably would if the issue was apparmor blocking access to some resources

seb128 commented:
Trying to use amidi -l in the snap env is giving

```
$ amidi -l
Dir Device    Name
ALSA lib conf.c:4553:(snd_config_update_r) Cannot access file /usr/share/alsa/alsa.conf
ALSA lib control.c:1528:(snd_ctl_open_noupdate) Invalid CTL hw:0
cannot open control for card 0: No such file or directory
ALSA lib conf.c:4553:(snd_config_update_r) Cannot access file /usr/share/alsa/alsa.conf
ALSA lib control.c:1528:(snd_ctl_open_noupdate) Invalid CTL hw:1
cannot open control for card 1: No such file or directory
```

lissyx commented:
I believe there is more context in https://forum.snapcraft.io/t/the-alsa-interface/7766/7

ebababi committed:
Exposed /usr/share/alsa

For ALSA to work, the $SNAP/usr/share/alsa needs to be exposed at /usr/share/alsa of the snap’s runtime by using the layout bind mechanism.

Reference: https://forum.snapcraft.io/t/the-alsa-interface/7766/7

ebababi commented:
@lissyx and @seb128, thank you for looking into this.


> Could you gives details on 'requires ALSA'? What exactly in alsa? Does it call into command line utilities?

My understanding is that the WebMIDI API implementation [wraps](https://searchfox.org/mozilla-central/source/dom/midi/midir_impl/src/lib.rs#276) and [uses](https://searchfox.org/mozilla-central/source/dom/midi/midirMIDIPlatformService.cpp#95) the [midr library](https://searchfox.org/mozilla-central/source/third_party/rust/midir) under the hood which, on Linux, [use ALSA](https://searchfox.org/mozilla-central/source/third_party/rust/midir/src/backend/alsa/mod.rs#9-10) to work with MIDI. It doesn't call into command line utilities.

> I've rebuilt a snap with the proposed change and it doesn't make a difference (trying the example code from https://www.smashingmagazine.com/2018/03/web-midi-api/), also the journal doesn't show any apparmor denial which it probably would if the issue was apparmor blocking access to some resources

An example site is https://webmidi-examples.glitch.me/ which, if snap allows, should display the detected MIDI devices, including the virtual ones, in the 'Devices' section. Take also a look at the browser console which, if no devices are detected, should write "Silently denying site request for MIDI access because no devices were detected. You may need to restart your browser after connecting a new device." as a result of [this code](https://searchfox.org/mozilla-central/source/dom/midi/MIDIPermissionRequest.cpp#162-169).

> Trying to use amidi -l in the snap env is giving [...]
> I believe there is more context in https://forum.snapcraft.io/t/the-alsa-interface/7766/7

Thank you; as written in the aforementioned link, with dd39a6a I have exposed at `/usr/share/alsa` of the snap’s runtime the `$SNAP/usr/share/alsa`. Can you please try this again, @lissyx?

seb128 commented:
Thanks, you need to stage libasound and include the /usr/share/alsa content in the snap for the content to be available, I added those changes and tested locally. It seems to not be enough for firefox to work but at least `amidi -l` works then in the snap env which is a first step, I proposed that now for review on https://github.com/canonical/firefox-snap/pull/37 against nightly

seb128 commented:
In fact it seems to be working now when trying on the debug console while on a website instead of the startpage

ebababi commented:
Closing in favor of #37.
